### ISSUE TITLE
Limit the "intrusion" of physics packages into albany main libs

### ISSUE THREAD

**Issue by bartgol** [SNLComputation/Albany#495]
I have been thinking about this issue (in different forms) for quite some time, and I think it impacts (to different extents) the productivity of every developer.

The issue: making (important) modifications/additions to the source/build files can be complicated by the fact that physics packages inject some of their code into the main albany files. Sometimes this seems unavoidable, but I believe we can at least do better.

My ideal situation would be to have the main albany folders (src, src/adapt, src/disc, src/evaluators, src/problems, src/responses, and src/utility) to be 'physics-free'. Of course, Albany _needs_ to be physics-packages aware somewhere. My proposal is to create a `factories` class, where we put all factories for all objects that require physics-package knowledge. For instance, problem factory, bc traits, discretization, and whatever else we can think of (analytic functions?).

This folder would be its own library, responsible of linking (in the CS but also figurative sense) Albany to the different packages.

What would the benefit be? Well, all but one folder of the main albany stuff would be physics free, they would contain "general" capabilities. There are some capabilities that, as of now, are ifdeffed out if a physics is not present, but can be thought as "general", so we could include them in the general code path, provided that they do not measurably impact performance/clarity of the code in the case that physics is not present. Of course, when adding capabilities to the main files, one needs to carefully think if such capabilities are "general" (i.e., could another package ever need it?) or "really package-specific". In the latter, we should strongly consider wrapping the functionality in an object that can be created via factory. In the first case, no need to ifdef it, even if no other package currently use it; and if they have a performance impact, we can ask the problem (or any other object that is created via a factory, hence being problem-specific) to tell us whether that operation needs to be performed, so that other physics that currently don't need it are not impacted.
Example: I took a look to `Albany_Application.cpp`, which is arguably the largest file we have (after `Albany_STKDiscretization.cpp` probably). There are quite a few LCM ifdeffed pieces of code, but most of them could be easily added to the main path (they would be either innocuous or not executed at all, such as some functions that only LCM solver call). The only exception is an extra scatter to set the overlapped residual back in the workset, which could be easily executed conditionally to the problem requesting it.

In short, I would really like to have an LCM/LandIce/Aeras/ATO/...-free main folder, and I think the code would be much clearer if all the physics-dependent parts were wrapped in a factory-based design.

I was wondering if anyone had some thoughts about this.

**Comment by bartgol** [comment]
I think the removal of all but LandIce has pretty much solved this.
